### PR TITLE
panda: refactor can_send

### PR DESCRIPTION
**Pull request by deanlee**

### CODE REVIEW

**Comment by briskspirit** [comment]
Not sure that this refactoring is needed or if it makes code more readable and clear. Now it is even more hard to understand pack_buffer function imo.

**Comment by gregjhogan** [comment]
I think there is an opportunity to decrease the complexity of `Panda::pack_can_buffer` while increasing readability (even if this isn't the best approach)

**Comment by briskspirit** [comment]
I am not sure that readability was increased.

**Comment by deanlee** [comment]
There should be a better way, I will try to figure it out.

**Comment by deanlee** [comment]
how about this one?

File: selfdrive/boardd/panda.cc [hunk 1]
```diff
@@ -363,57 +363,47 @@ uint8_t Panda::len_to_dlc(uint8_t len) {
...
 }
 
-void Panda::pack_can_buffer(const capnp::List<cereal::CanData>::Reader &can_data_list,
-                         std::function<void(uint8_t *, size_t)> write_func) {
-  if (send.size() < (can_data_list.size() * CANPACKET_MAX_SIZE)) {
-    send.resize(can_data_list.size() * CANPACKET_MAX_SIZE);
+static void write_packet(uint8_t *dest, int *write_pos, const uint8_t *src, size_t size) {
+  for (int i = 0, &pos = *write_pos; i < size; ++i, ++pos) {
+    if (pos % USBPACKET_MAX_SIZE == 0) {
+      dest[pos++] = pos / USBPACKET_MAX_SIZE;
```
Comment (briskspirit) [inline selfdrive/boardd/panda.cc:369]:
```suggestion
      dest[pos] = pos / USBPACKET_MAX_SIZE;
      pos++;
```
This is MISRA violation. We are not using MISRA for OP, but I want code related to panda be as close as possible to MISRA rules.

File: selfdrive/boardd/panda.cc [hunk 2]
```diff
@@ -363,57 +363,48 @@ uint8_t Panda::len_to_dlc(uint8_t len) {
...
+
+    write_packet(send_buf, &pos, (uint8_t *)&header, sizeof(can_header));
+    write_packet(send_buf, &pos, (uint8_t *)can_data.begin(), can_data.size());
+    if (pos >= USB_TX_SOFT_LIMIT) {
+      write_func(send_buf, pos);
+      pos = 0;
     }
   }
+
+  // flush the remain packets
```
Comment (briskspirit) [inline selfdrive/boardd/panda.cc:406]:
```suggestion
  // send remaining packets
```

File: selfdrive/boardd/panda.cc [hunk 3]
```diff
@@ -363,57 +363,48 @@ uint8_t Panda::len_to_dlc(uint8_t len) {
   }
 }
 
-void Panda::pack_can_buffer(const capnp::List<cereal::CanData>::Reader &can_data_list,
-                         std::function<void(uint8_t *, size_t)> write_func) {
-  if (send.size() < (can_data_list.size() * CANPACKET_MAX_SIZE)) {
-    send.resize(can_data_list.size() * CANPACKET_MAX_SIZE);
+static void write_packet(uint8_t *dest, int *write_pos, const uint8_t *src, size_t size) {
+  for (int i = 0, &pos = *write_pos; i < size; ++i, ++pos) {
+    if (pos % USBPACKET_MAX_SIZE == 0) {
```
Comment (briskspirit) [inline selfdrive/boardd/panda.cc:368]:
```suggestion
    // Insert counter every 64 bytes (first byte of 64 bytes USB packet)
    if (pos % USBPACKET_MAX_SIZE == 0) {
```

**Pull request merged**